[ENH] Transformers module full refactor - part I, `series` module

fkiraly opened the pull request:
This PR refactors all transformers in the `transformations` module to the new interface.
Conditional on #1790 and #1774 which are included.

This is part I - the entire `transformations.series` module, with exception of STLtransformer which is left untouched due to current discussions around the transformer (see #1677).

For all transformers, the following was carried out:
* inheritance parent is now `BaseTransformer`
* adding tags as appropriate
* moving logic to underscore methods `_fit`, `_transform`, and `_inverse_transform` (where applicable)
* removal of check/conversion logic which is already covered by `BaseTransformer`
* linting and cleaning up docstrings

Reviewers should focus on individual transformers and the general pattern.

Changes to extension templates should be reviewed in #1774, not here.
Changes to `BaseTransformer` should be reviewed in #1790, not here.

fkiraly commented:
@TonyBagnall, FYI. This is transformers, so also closely linked with TSC.
The refactored transformers should work out-of-the box on `numpy3D` inputs (they didn't before).

fkiraly commented:
@lmmentel, would also appreciate a look and comments, since you recently looked into the transformers tutorial #1705, and the relevance for `FeatureUnion`.

GuzalBulatova reviewed (commented):
Great work (both figuratively and literally)! I appreciate the resulting uniformity.

chrisholder commented on sktime/transformations/series/difference.py:
```diff
@@ -161,60 +169,70 @@ def _check_inverse_transform_index(self, Z):
+        """
+        Fit transformer to X and y.
+
+        core logic
```
Ok im not going to mark anymore of these 'core logic' comments

fkiraly replied:
same

fkiraly commented on sktime/transformations/series/adapt.py:
```diff
@@ -60,64 +39,101 @@ class TabularToSeriesAdaptor(_SeriesToSeriesTransformer):
-        Z = check_series(Z)
-        Zt = self.transformer_.inverse_transform(_from_series_to_2d_numpy(Z))
-        return _from_2d_numpy_to_series(Zt, index=Z.index)
+        from sklearn.preprocessing import StandardScaler
```
yes, the specific reason is that the import is needed only in testing (when `get_test_params` is called) hence can be completely ignored in usage.

fkiraly commented:
Re "core logic", this should perhaps be better "implementation of abstract method which contains core logic"?

fkiraly commented on sktime/transformations/series/boxcox.py:
```diff
@@ -111,69 +115,77 @@ def __init__(self, bounds=None, method="mle", sp=None):
-        zt = boxcox(z.to_numpy(), self.lambda_)
-        return pd.Series(zt, index=z.index)
+        X_shape = X.shape
+        Xt = boxcox(X.flatten(), self.lambda_)
```
hm, yes, you're right.
I'm not sure anymore why I changed this.

Question: would it make more sense to do this by variable then? Or should users apply the `ColumnTransformer` instead?
Question: would it make sense later to have a transformer where `lambda` can be input?

fkiraly replied:
I've now set the `univariate-only` tag to `True`. That should allow only n x 1 arrays (i.e., univariate series in the numpy format), and all should be fine?

GuzalBulatova replied:
Thanks! yes this should work fine as now the length will be addressed as intended 👍 
Q1: I think it would be more user-friendly if we add functionality to handle multivariate input here, either by calling ColumnTransformer from BoxCox or adding the variable-wise transform. Don't know how far we'd like to go with complicating it though, in terms of design it seems to be a bit excessive. My vote would be that even though it's easier for a user to have package deals, we should leave it modular.
Q2: It depends whether there's a scenario this transformation will be used for some other reasons rather than approximating the data to normal distribution. I don't think it's necessary to enable input here, but maybe you have some use cases in mind?
An enhancement that we can add here might be adding [Yeo-Johnson method](https://www.jstor.org/stable/2673623), enabling negative values. What do you think?

fkiraly replied:
Q1: yes, I've been thinking about that. Currently, my favourite solution would be "apply-by-variable" next to "apply-by-instance", i.e., univariate forecasters/transformers are cast across variables, just like simple forecasters are cast across instances in a panel. That may require some additional work, and in most cases a direct implemention inside the transformer would be more efficient.

Q2: hm, not sure - generally, I like to not make too many assumptions on how things will or have to be used, since often there are perfectly sensible application cases outside of what one can think of individually. Regarding Yeo-Johnson, this sounds like a sensible extension? Would allow to cover negative values, and up to the strategy of how lambda is determined, it is the same as BoxCox in the case we currently cover (all values positive)?

fkiraly commented on sktime/transformations/series/theta.py:
```diff
@@ -62,6 +61,14 @@ class ThetaLinesTransformer(_SeriesToSeriesTransformer):
     _tags = {
+        "scitype:transform-input": "Series",
+        # what is the scitype of X: Series, or Panel
+        "scitype:transform-output": "Series",
```
that's still a `Series`, it just has multiple variables.
(at least per `sktime` mtype specification).

fkiraly commented on sktime/transformations/series/theta.py:
```diff
@@ -71,24 +78,25 @@ def __init__(self, theta=(0, 2)):
-        theta_lines: ndarray or pd.DataFrame
+        theta_lines: pd.Series or pd.DataFrame
             Transformed series: single Theta-line or a pd.DataFrame of
             shape: len(Z)*len(self.theta).
```
good catch

fkiraly replied:
updated; I also made further updates to the docstring (hope they are accurate)

fkiraly commented on sktime/transformations/series/adapt.py:
```diff
@@ -60,64 +39,101 @@ class TabularToSeriesAdaptor(_SeriesToSeriesTransformer):
+
+        Returns
+        -------
+        inverse transformed version of X
```
this depends on #1790 and you already flagged it there. I've merged my proposed changes into here.
Let's discuss any potential additional changes in #1790.

fkiraly commented on sktime/transformations/series/boxcox.py:
```diff
@@ -204,45 +216,59 @@ class LogTransformer(_SeriesToSeriesTransformer):
+    def _transform(self, X, y=None):
+        """Transform X and return a transformed version.
+
+        core logic
```
changing this consistently to "private _fit containing the core logic, called from fit", as discussed on slack

fkiraly commented on sktime/transformations/series/outlier_detection.py:
```diff
@@ -6,16 +6,16 @@
 __author__ = ["Martin Walter"]
```
thanks, changed

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -80,73 +89,105 @@ def __init__(self, transformer, passthrough=False):
+    def _fit(self, X, y=None):
+        """Fit transformer to X and y.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -80,73 +89,105 @@ def __init__(self, transformer, passthrough=False):
+    def _transform(self, X, y=None):
+        """Transform X and return a transformed version.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -80,73 +89,105 @@ def __init__(self, transformer, passthrough=False):
+    def _inverse_transform(self, X, y=None):
+        """Inverse transform, inverse operation to transform.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -80,73 +89,105 @@ def __init__(self, transformer, passthrough=False):
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        from sktime.transformations.series.boxcox import BoxCoxTransformer
```
imports for testing should be in here, since they are not needed in actual usage

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -184,34 +225,51 @@ class ColumnwiseTransformer(_SeriesToSeriesTransformer):
 
-        Iterates over columns (series) and applies
-        the fit function of the transformer.
+        core logic
```
same

fkiraly commented on sktime/transformations/series/compose.py:
```diff
@@ -339,6 +384,22 @@ def update(self, Z, X=None, update_params=True):
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        from sktime.transformations.series.detrend import Detrender
```
same

fkiraly commented on sktime/transformations/series/cos.py:
```diff
@@ -30,29 +29,57 @@ class CosineTransformer(_SeriesToSeriesTransformer):
+    def _transform(self, X, y=None):
+        """Transform X and return a transformed version.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/date.py:
```diff
@@ -105,28 +113,28 @@ def __init__(self, ts_freq=None, feature_scope="minimal", manual_selection=None)
+        """Transform X and return a transformed version.
 
-        Returns a transformed version of Z.
+        core logic
```
same

fkiraly commented on sktime/transformations/series/detrend/_deseasonalize.py:
```diff
@@ -73,121 +85,125 @@ def __init__(self, sp=1, model="additive"):
+        """Transform X and return a transformed version.
 
-        Returns a transformed version of y.
+        core logic
```
same

fkiraly commented on sktime/transformations/series/detrend/_detrend.py:
```diff
@@ -65,29 +64,44 @@ class Detrender(_SeriesToSeriesTransformer):
+    def _fit(self, X, y=None):
+        """Fit transformer to X and y.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/detrend/_detrend.py:
```diff
@@ -96,31 +110,31 @@ def fit(self, Z, X=None):
+    def _transform(self, X, y=None):
+        """Transform X and return a transformed version.
+
+        core logic
```
same

fkiraly commented on sktime/transformations/series/detrend/_detrend.py:
```diff
@@ -184,37 +199,55 @@ def inverse_transform(self, Z, X=None):
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        from sktime.forecasting.exp_smoothing import ExponentialSmoothing
```
same

fkiraly commented:
@chrisholder, I think I addressed/commented on your change requests:

* as discussed on slack, I replaced the slightly obscure "core logic" line in the docstrings with something more extensive, throughout the transformer module
* imports in `get_test_params` should stay in there, since we're using them only in testing. By design, when the package is used (not tested), the method can be ignored, so it should stay "hidden".

fkiraly commented on sktime/transformations/base.py:
```diff
@@ -560,11 +436,387 @@ def fit_transform(self, X, y=None, Z=None):
+            )
+        else:
+            # output_scitype is "Panel" and no need for conversion
+            pass
```
ok, done, and clarified the comment so it's informative what the "else" case would be

fkiraly commented:
@GuzalBulatova, I think I addressed your comments?
* `pandas.DataFrame` is interpreted as a multivariate series (not as a univariate panel), so the first thing is fine
* in the `BoxCoxTransformer`, I've set the `univariate-only` tag to `True`, so it's an 1:1 mirror of pre-refactor function now

fkiraly commented:
Merging since I would like to use this as example to point to in another PR, and @GuzalBulatova's review was a comment not a change request, so formally not a blocker.
@GuzalBulatova, kindly let me know (or unmerge) if you feel there is still outstanding items on Theta - I thought I've addressed them by reverting.
